stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+#ifndef STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+#define STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/row.hpp>
+#include <stan/math/prim/fun/col.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/exp.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <stan/math/prim/core.hpp>
+#include <Eigen/Core>
+#include <vector>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * pi(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass.
+ * The [in, out] argument are saved so that we can use them when
+ * calculating the derivatives.
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @param[in, out] alphas unnormalized partial marginal density.
+ *                   The jth column is the joint density over all
+ *                   observations y and the hidden state j.
+ * @param[in, out] alpha_log_norms max coefficient for column of alpha,
+ *                   to be used to normalize alphas.
+ * @param[in, out] omegas term-wise exponential of omegas.
+ * @return log marginal density.
+ */
+double hmm_marginal_lpdf(const Eigen::MatrixXd& log_omegas,
+                         const Eigen::MatrixXd& Gamma,
+                         const Eigen::VectorXd& rho, Eigen::MatrixXd& alphas,
+                         Eigen::VectorXd& alpha_log_norms,
+                         Eigen::MatrixXd& omegas) {
+  omegas = log_omegas.array().exp();  // CHECK -- why the .array()?
+  int n_states = log_omegas.rows();
+  int n_transitions = log_omegas.cols() - 1;
+
+  alphas.col(0) = omegas.col(0).cwiseProduct(rho);
+
+  double norm = alphas.col(0).maxCoeff();
+  alphas.col(0) /= norm;
+  alpha_log_norms(0) = std::log(norm);
+
+  for (int n = 0; n < n_transitions; ++n) {
+    alphas.col(n + 1)
+        = omegas.col(n + 1).cwiseProduct(Gamma.transpose() * alphas.col(n));
+
+    double norm = alphas.col(n + 1).maxCoeff();
+    alphas.col(n + 1) /= norm;
+    alpha_log_norms(n + 1) = std::log(norm) + alpha_log_norms(n);
+  }
+
+  return log(alphas.col(n_transitions).sum()) + alpha_log_norms(n_transitions);
+}
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
```

Comment:
This should not have "marginal" in its name because we think of the latent states as parameters and thus they are naturally marginalized out.  In the traditional literature, they distinguished between the "data density" (for y) and the "complete data distribution" (for y and the latent states z), because strict frequentists have to pretend the mixture responsibility parameters are data so that they don't have a philosophical meltdown.  You see that usage all over the EM literature, where the latent states z get marginalized out.

Reply:
I'll ask for a second opinion on this (@vianeylb, @betanalpha). As someone familiar with a Bayesian framework and a candid approach to the problem, this is clearly a marginal distribution. Naturally, I'm willing to accomodate the user.

Reply:
In the Cappé et al book, the distribution of the observations  is referred to as the "marginal distribution of the observations only", obtained through the process of marginalization of the state variables. 

In the Zucchini et al book, they have a section on marginal distributions of a single observation, Y_t, and discuss higher order marginal distributions (e.g. [Y_t, Y_t+k]). Extending this, the distribution of the data only also falls under what they refer to as marginal distributions. 

The likelihood in both contexts comes from what they refer to as the marginal distribution of the data only, and they emphasize that this quantity is obtained through marginalization over the state variables. Given that inference for HMMs via the likelihood (marginal distribution of the data) vs the complete-data likelihood seems to be a topic that people are highly opinionated about, emphasizing that inference for the parameters of an HMM in Stan is done through marginalization over the state variables and use of the marginal distribution of the data directly seems preferable to me. And as the Zucchini book is clearly more on the frequentist side, I don't think the "marginal_lpdf" will be divisive to use. Just my opinion.

Reply:
I agree with Vianey!

Reply:
We all agree that what's at play is a joint Bayesian model `p(y, z, theta)` where `y` is observed data, `z` is unobserved, and `theta` are parameters.  

The terminology I'm familiar with calls `p(y | theta)` the "likelihood" and `p(y, z | theta)` the "full data likelihood", because frequentist philosophy melts down if `z` is considered a parameter.

I looked up Zucchini et al. and in section 2.3, they use likelihood in exactly this way for `p(y | theta)` with the states `z` marginalized out of `p(y, z | theta)`.    They then talk about marginalizing `p(y | theta)` to subsequences of `y`.  In general, if we have a joint distribution `p(y)`, we don't talk about `p(y)` as a marginal distribution even though it is the degenerate case with nothing marginalized out.  That's why I find calling `p(y | theta)` a marginal distribution confusing in this context.

I'd also strongly prefer not to have verbose names for things if we're not going to be distinguishing among many alternatives.